src/stream/xml_writer.rs

```diff
@@ -157,81 +144,125 @@ impl<W: Write> XmlWriter<W> {
     fn write_start_array(&mut self, _len: Option<u64>) -> Result<(), Error> {
-        self.write_value_event(EventKind::StartArray, |this| {
-            this.start_element("array")?;
-            this.stack.push(Element::Array);
-            Ok(())
-        })
+        self.handle_pending_collection()?;
+        self.pending_collection = Some(PendingCollection::Array);
+        Ok(())
     }
 
     fn write_start_dictionary(&mut self, _len: Option<u64>) -> Result<(), Error> {
-        self.write_value_event(EventKind::StartDictionary, |this| {
-            this.start_element("dict")?;
-            this.stack.push(Element::Dictionary);
-            Ok(())
-        })
+        self.handle_pending_collection()?;
+        self.pending_collection = Some(PendingCollection::Dictionary);
+        Ok(())
     }
 
     fn write_end_collection(&mut self) -> Result<(), Error> {
         self.write_event(|this| {
-            match (this.stack.pop(), this.expecting_key) {
-                (Some(Element::Dictionary), true) => {
+            match (
+                this.stack.pop(),
+                this.expecting_key,
+                this.pending_collection.as_ref(),
+            ) {
+                (element, _, Some(&PendingCollection::Dictionary)) => {
+                    if let Some(element) = element {
+                        this.stack.push(element)
+                    }; // We didn't really want this
+                    this.xml_writer
+                        .write_event(XmlEvent::Empty(BytesStart::borrowed_name(b"dict")))?;
+                }
+                (Some(Element::Dictionary), true, None) => {
                     this.end_element("dict")?;
                 }
-                (Some(Element::Array), _) => {
+                (element, _, Some(&PendingCollection::Array)) => {
+                    if let Some(element) = element {
+                        this.stack.push(element)
+                    }; // We didn't really want this
+                    this.xml_writer
+                        .write_event(XmlEvent::Empty(BytesStart::borrowed_name(b"array")))?;
+                }
+                (Some(Element::Array), _, None) => {
                     this.end_element("array")?;
                 }
-                (Some(Element::Dictionary), false) | (None, _) => {
+                (Some(Element::Dictionary), false, _) | (None, _, _) => {
                     return Err(ErrorKind::UnexpectedEventType {
                         expected: EventKind::ValueOrStartCollection,
                         found: EventKind::EndCollection,
                     }
                     .without_position());
                 }
             }
+            this.pending_collection = None;
             this.expecting_key = this.stack.last() == Some(&Element::Dictionary);
             Ok(())
         })
     }
 
     fn write_boolean(&mut self, value: bool) -> Result<(), Error> {
+        self.handle_pending_collection()?;
```

Comment:
Can this go in `write_value_event` so it's not duplicated across all the functions below?

Reply:
That needed a bit of restructuring, but yeah, good call.